TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+//          { "type" : [ "string", "object", "null" ] }
+- (NSSet *)parseTypeKeywordFromSchema:(NSDictionary *)schema
+{
+    id typeValue = schema[kTWTJSONSchemaKeywordType];
+
+    if (!typeValue) {
+        return nil;
+    }
+
+    // Relies on parseSchema: having filtered out schemas with type array, object, string, number, or integer. Thus, type is either:
+    // A) One of these strings: "boolean", "null", "any"
+    // B) An array of valid type strings
+    // C) Invalid wrt the JSON Schema definition of the type keyword
+
+    [self pushPathComponent:kTWTJSONSchemaKeywordType];
+
+    static NSSet *allowedTypes = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        allowedTypes = [NSSet setWithObjects:kTWTJSONSchemaTypeKeywordAny, kTWTJSONSchemaTypeKeywordBoolean, kTWTJSONSchemaTypeKeywordNull, nil];
+    });
+
+    if ([allowedTypes containsObject:typeValue]) {
+        [self popPathComponent];
+        return [NSSet setWithObject:typeValue];
+    }
+
+    [self checkObjectIsArrayWithAtLeastOneItem:typeValue allowNil:NO];
+
+    [typeValue enumerateObjectsUsingBlock:^(NSString *typeString, NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [self checkObject:typeString isOneOf:[self validJSONTypeKeywords]];
+        [self popPathComponent];
+    }];
+
+    [self popPathComponent];
+    return [NSSet setWithArray:typeValue];
+}
+
+
+// Keywords: exclusiveMinimum & Maximum, uniqueItems
+- (BOOL)parseBooleanValueFromSchema:(NSDictionary *)schema forKey:(NSString *)key valueIfNotPresent:(BOOL)defaultValue
+{
+    [self pushPathComponent:key];
+    NSNumber *booleanObject = schema[key];
+    [self checkObject:booleanObject isOfClass:[NSNumber class] allowNil:YES];
+    [self popPathComponent];
+    return booleanObject ? [booleanObject boolValue] : defaultValue;
+}
+
+
+// Keywords: pattern
+- (NSString *)parseStringFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSString *string = schema[key];
+    [self checkObject:string isOfClass:[NSString class] allowNil:YES];
+    [self popPathComponent];
+    return string;
+}
+
+
+// Keywords: minimum, maximum
+- (NSNumber *)parseNumberFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSNumber *number = schema[key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+    [self popPathComponent];
+    return number;
+}
+
+
+// Keywords: multipleOf
+- (NSNumber *)parsePositiveNumberFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    NSNumber *number = schema[key];
+    if (!number) {
+        return nil;
+    }
+
+    [self pushPathComponent:key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+    if ([number doubleValue] <= 0) {
+        NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected number greater than zero.", key];
```

Comment:
I wonder if we can’t get rid of the “Invalid value for key %@” bit. If we generate error messages in the right way, they’ll say something like: “Error at /a/b/c/multipleOf: expected positive number, but got %@", number.